File: dir_rangan_playhouse/interp_p_to_q.m
function S_q_ = interp_p_to_q(n_r,n_w_,n_A,S_p_);

str_thisfunction = 'interp_p_to_q';

if (nargin<4);
test_transforms_dr.m();
disp('returning'); return;
end;%if (nargin<4);

n_S = idivide(cast(numel(S_p_),'int64'),cast(n_A,'int64'));
if numel(S_p_)~=n_A*n_S;
disp(sprintf(' %% Warning, n_A %d n_S %d in %s',n_A,n_S,str_thisfunction));
end;%if numel(S_p_)~=n_A*n_S;

%%%%%%%%%%%%%%%%;
if (numel(unique(n_w_))> 1);
S_q_ = zeros(n_A*n_S,1);
%%%%%%%%;
if n_S==1;
%%%%;
% single transformation on adaptive grid. ;
%%%%;
n_w_max = n_w_(1+n_r-1);
ic=0;
for nr=0:n_r-1;
n_w = n_w_(1+nr);
if (n_w>0);
tmp_index_ = ic + [0:n_w-1];
S_q_(1+tmp_index_) = fft(S_p_(1+tmp_index_))/max(1,sqrt(n_w));
ic = ic + n_w;
end;%if (n_w>0);
end;%for nr=0:n_r-1;
assert(ic==n_A);
%%%%;
end;%if n_S==1;
%%%%%%%%;
if n_S> 1;
for nS=0:n_S-1;
tmp_index_ = nS*n_A + [0:n_A-1];
S_q_(1+tmp_index_) = interp_p_to_q(n_r,n_w_,n_A,S_p_(1+tmp_index_));
end;%for nS=0:n_S-1;
S_q_ = reshape(S_q_,size(S_p_));
end;%if n_S> 1;
%%%%%%%%;
end;%if (numel(unique(n_w_))> 1);
%%%%%%%%%%%%%%%%;

%%%%%%%%%%%%%%%%;
if (numel(unique(n_w_))==1);
n_w = n_w_(1+0);
S_q_ = reshape(fft(reshape(S_p_,[n_w,n_r,n_S]),[],1+0)/max(1,sqrt(n_w)),[numel(S_p_),1]);
S_q_ = reshape(S_q_,size(S_p_));
end;%if (numel(unique(n_w_))==1);
%%%%%%%%%%%%%%%%;

File: dir_rangan_python/interp_p_to_q.py
from dir_matlab_macros import * ;

def interp_p_to_q(
        n_r=None,
        n_w_=None,
        n_A=None,
        S_p_=None,
):
    flag_verbose = 0;
    str_thisfunction = 'interp_p_to_q' ;

    #%%%%%%%%;
    if (flag_verbose): print(f' %% [entering {str_thisfunction}]');
    #%%%%%%%%;

    n_S = int( numel(S_p_) / int(n_A) );
    if numel(S_p_)!=n_A*n_S: print(f' %% Warning, n_A {n_A} n_S {n_S} in {str_thisfunction}');

    #%%%%%%%%%%%%%%%%;
    if numel_unique(n_w_) > 1:
        S_q_ = torch.zeros(n_A*n_S).to(dtype=torch.complex64);
        #%%%%%%%%;
        if n_S==1:
            #%%%%;
            #% single transformation on adaptive grid. ;
            #%%%%;
            n_w_max = int(n_w_[n_r-1].item());
            ic=0;
            for nr in range(n_r):
                n_w = int(n_w_[nr]).item();
                if (n_w>0):
                    tmp_index_ = int(ic) + torch.arange(n_w).to(dtype=torch.int32);
                    S_q_[tmp_index_] = torch.tensor(np.fft.fft(S_p_[tmp_index_].numpy())/np.maximum(1,np.sqrt(n_w))).to(dtype=torch.complex64);
                    ic = ic + n_w;
                #end;%if (n_w>0);
            #end;%for nr=0:n_r-1;
            assert(ic==n_A);
            #%%%%;
        #end;%if n_S==1;
        #%%%%%%%%;
        if n_S> 1:
            for nS in range(n_S):
                tmp_index_ = int(nS*n_A) + torch.arange(n_A).to(dtype=torch.int32);
                S_q_[tmp_index_] = interp_p_to_q(n_r,n_w_,n_A,S_p_[tmp_index_]);
            #end;%for nS=0:n_S-1;
        #end;%if n_S> 1;
        #%%%%%%%%;
    #end;%if (numel(unique(n_w_))> 1);
    #%%%%%%%%%%%%%%%%;

    #%%%%%%%%%%%%%%%%;
    if numel_unique(n_w_)==1:
        n_w = int(n_w_[0].item());
        S_q_ = torch.fft.fft(torch.reshape(S_p_,mtr((n_w,n_r,n_S))),axis=2-0).to(dtype=torch.complex64).ravel() / np.maximum(1,np.sqrt(n_w));
        assert(numel(S_q_)==numel(S_p_));
    #end;%if (numel(unique(n_w_))==1);
    #%%%%%%%%%%%%%%%%;

    #%%%%%%%%;
    if (flag_verbose): print(f' %% [finished {str_thisfunction}]');
    #%%%%%%%%;
    return(S_q_);


